### DISCUSSION TITLE
Usage with AWS Lambda

### DISCUSSION THREAD

**Discussion by jamesopstad** [gajus/slonik#520] [Q&A]
What is the recommended way to use this library with AWS Lambda? I am connecting to PgBouncer, which is running in transaction mode. I've run into some problems using the `pg` library directly in the past with connections not being closed.

* Should the pool be created inside or outside the handler function?
* Does `pool.end()` need to be explicitly called?
* Does the pool need to be configured in a particular way?

Thanks in advance.

**Comment by firxworx** [comment]
For code that may or may not run in lambda I define a boolean `const IS_AWS_LAMBDA_ENVIRONMENT = !!process.env.AWS_LAMBDA_FUNCTION_NAME` to use in conditionals.

In pool config I have used `maximumPoolSize: IS_AWS_LAMBDA_ENVIRONMENT ? 1 : 10` (slonik's default is 10)

If you are using slonik as documented in the README the question of where and how it handles equivalent of `.end()` for you is covered. The maintainer/founder gajus has also posted some Medium articles and posts over the years which are extremely illuminating on how slonik works and to provide examples (careful when reading -- he sometimes uses naming that could refer to more than one thing).

The inside or outside question is a nuanced one imho because of how Lambda and other FaaS environments can keep variables defined outside the handler alive between invocations for the duration of the environment. The Prisma docs have some background on this https://www.prisma.io/docs/orm/prisma-client/setup-and-configuration/databases-connections#serverless-environments-faas and I recall AWS has some articles as well. 

With serverless you'd definitely want to use a connection pooler like RDS Proxy or PgBouncer to avoid exhausting your DB under load.
